der/src/document.rs

```diff
@@ -1,67 +1,102 @@
 //! ASN.1 DER-encoded documents stored on the heap.
 
 use crate::{Decodable, Encodable, Error, Result};
-use alloc::{boxed::Box, vec::Vec};
+use alloc::vec::Vec;
+use core::marker::PhantomData;
 
 #[cfg(feature = "pem")]
-use {crate::pem, alloc::string::String};
+use {
+    crate::pem::{self, PemLabel},
+    alloc::string::String,
+};
 
 #[cfg(feature = "std")]
 use std::{fs, path::Path};
 
+/// Marker trait which identifies documents which contain sensitive
+/// information, such as private keys.
+pub trait Sensitivity {
+    /// Does this type contain potentially sensitive data?
+    ///
+    /// This enables hardened file permissions when persisting data to disk.
+    const SENSITIVE: bool;
+}
+
 /// ASN.1 DER-encoded document.
 ///
-/// This trait is intended to impl on types which contain an ASN.1 DER-encoded
-/// document which is guaranteed to encode as the associated `Message` type.
+/// This type wraps an encoded ASN.1 DER message which is guaranteed to
+/// infallibly decode as type `T`.
 ///
 /// It implements common functionality related to encoding/decoding such
 /// documents, such as PEM encapsulation as well as reading/writing documents
 /// from/to the filesystem.
 #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
-pub trait Document<'a>: AsRef<[u8]> + Sized + TryFrom<Vec<u8>, Error = Error> {
-    /// ASN.1 message type this document decodes to.
-    type Message: Decodable<'a> + Encodable + Sized;
+pub struct Document<T: Sensitivity> {
+    /// ASN.1 DER-encoded document guaranteed to decode to `T` infallibly.
+    der_bytes: Vec<u8>,
 
-    /// Does this type contain potentially sensitive data?
-    ///
-    /// This enables hardened file permissions when persisting data to disk.
-    const SENSITIVE: bool;
+    /// Rust type corresponding to the ASN.1 DER message the bytes can be
+    /// infallibly deserialized as.
+    msg_type: PhantomData<T>,
+}
 
-    /// Borrow the inner serialized bytes of this document.
-    fn as_der(&self) -> &[u8] {
-        self.as_ref()
-    }
+impl<T> Document<T> where T: Sensitivity {}
 
-    /// Return an allocated ASN.1 DER serialization as a boxed slice.
-    fn to_der(&self) -> Box<[u8]> {
-        self.as_ref().to_vec().into_boxed_slice()
+impl<T> Document<T>
+where
+    T: Sensitivity,
+{
+    /// Borrow the inner serialized bytes of this document.
+    pub fn as_der(&self) -> &[u8] {
+        self.der_bytes.as_slice()
     }
```

Comment:
We don't actually know if the bytes are `der` because we haven't decoded them yet. Perhaps we should just let `AsRef` or even `Deref` do the heavy lifting here?

Reply:
If you take a look at the definition of `from_der`, you'll see it checks the document decodes using `doc.try_decode()` before returning it to the caller. The type definitely does maintain the invariant that `self.der_bytes` will always decode successfully, and as I noted in #421, it may be possible to store the parsed message to prevent repeat decoding (definitely not going to attempt that in this PR).

Also for some reason GitHub isn't letting me reply to your comment about the PEM label, but there's already a `PemLabel` trait for that.

And FWIW locally I'm trying to replace the trait with a const generic parameter, and liking that better for now.